File: src/15_Spanning_Trees/prim.cpp
#include <bits/stdc++.h>
using namespace std;


const int m = 8;
vector<tuple<int, int, int> > edge_list = {
  make_tuple(1, 2, 3),
  make_tuple(1, 5, 5),
  make_tuple(2, 5, 6),
  make_tuple(2, 3, 5),
  make_tuple(5, 6, 2),
  make_tuple(3, 6, 3),
  make_tuple(3, 4, 9),
  make_tuple(6, 4, 7)
};

const int n = 12;
const int source = 1;
bool visited[n + 1];
int dist[n + 1];
int last[n + 1];
vector<pair<int, int> > mst;
// each pair: <head, edge_weight>
vector<pair<int, int> > adj[n + 1] = {
  {},
  {{2, 3}, {5, 5}},          //  1
  {{1, 3}, {3, 5}, {5, 6}},  //  2
  {{2, 5}, {4, 9}, {6, 3}},  //  3
  {{3, 9}, {6, 7}},          //  4
  {{1, 5}, {2, 6}, {6, 2}},  //  5
  {{3, 3}, {4, 7}, {5, 2}}   //  6
};
// Using greater as predicate, we get a min priority queue. Alternatively, the
// default max priority queue can be used (with less instead of greater), but
// the pair pushed upon relaxation should contain -dist[node], for the smaller
// distance to get higher priority.
// each pair: <dist[node], node>
priority_queue<pair<int, int>,
               vector<pair<int, int> >,
               greater<pair<int, int> > > pq;


// O((m + n)logn), n: number of nodes, m: number of edges
void prim(int s)
{
  memset(dist, 0x3f, sizeof(dist));
  dist[s] = 0;
  last[s] = -1;
  memset(visited, false, sizeof(visited));
  pq.push(make_pair(0, s));

  while (!pq.empty()) {
    int u = pq.top().second; pq.pop();
    if (visited[u]) continue;
    visited[u] = true;
    if (u != s) mst.push_back(make_pair(u, last[u]));
    for (auto& [v, w] : adj[u]) {
      if (dist[v] > w) {
        dist[v] = w;
        last[v] = u;
        pq.push(make_pair(dist[v], v));
      }
    }
  }
}


int main()
{
  prim(source);
  for (auto& [a, b] : mst) cout << a << " - " << b << '\n';
}

File: src/15_Spanning_Trees/prim.py
import math
import heapq

edge_list = [
  [1, 2, 3],
  [1, 5, 5],
  [2, 5, 6],
  [2, 3, 5],
  [5, 6, 2],
  [3, 6, 3],
  [3, 4, 9],
  [6, 4, 7]
]
m = len(edge_list)
adj = []


def populate_adj():
  for a, b, w in edge_list:
    max_limit = max(a, b)
    if len(adj) - 1 < max_limit:
      for _ in range(max_limit + 1 - len(adj)):
        adj.append([])
    adj[a].append([b, w])
    adj[b].append([a, w])


# O(n + mlog(n))
def prim(start):
  visited = [False] * len(adj)
  dist = [math.inf] * len(adj)
  prev = [None] * len(adj)
  dist[start] = 0
  prev[start] = -1
  # The priority queue at Dijkstra's alg keeps the min distance from start to
  # the node. At Prim's alg, it keeps the min edge length from the node to a
  # neighbor.
  pq = [(0, start)]
  heapq.heapify(pq)
  cost = 0

  while pq:
    w, u = heapq.heappop(pq)
    if visited[u]:
      continue
    visited[u] = True
    cost += w
    if u != start:
      print(prev[u], '-', u)
    for v, w in adj[u]:
      if w < dist[v]:
        dist[v] = w
        prev[v] = u
        heapq.heappush(pq, (w, v))

  print("cost:", cost)

if __name__ == '__main__':
  populate_adj()
  prim(1)
